File: CreationalPattern/BuilderPattern/BuilderPattern.cs
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Text;

/// <summary>
/// Builder for constructing an HttpRequestMessage object in a clean and fluent way.
/// </summary>
public class HttpRequestBuilder
{
    private HttpMethod method;
    private Uri uri;
    private HttpContent content;
    private readonly Dictionary<string, string> headers = new();

    /// <summary>
    /// Sets the HTTP method (GET, POST, etc.)
    /// </summary>
    public HttpRequestBuilder SetMethod(HttpMethod method)
    {
        this.method = method;
        return this;
    }

    /// <summary>
    /// Sets the URI for the request.
    /// </summary>
    public HttpRequestBuilder SetUri(string url)
    {
        this.uri = new Uri(url);
        return this;
    }

    /// <summary>
    /// Sets the content (body) of the request.
    /// </summary>
    public HttpRequestBuilder SetContent(string body, string mediaType = "application/json")
    {
        this.content = new StringContent(body, Encoding.UTF8, mediaType);
        return this;
    }

    /// <summary>
    /// Adds a header to the request.
    /// </summary>
    public HttpRequestBuilder AddHeader(string key, string value)
    {
        headers[key] = value;
        return this;
    }

    /// <summary>
    /// Builds the HttpRequestMessage instance.
    /// </summary>
    public HttpRequestMessage Build()
    {
        var request = new HttpRequestMessage(method, uri);
        if (content != null) request.Content = content;
        foreach (var header in headers)
            request.Headers.TryAddWithoutValidation(header.Key, header.Value);

        return request;
    }
}

// Usage
public class Program
{
    public static void Main()
    {
        var request = new HttpRequestBuilder()
            .SetMethod(HttpMethod.Post)
            .SetUri("https://api.example.com/users")
            .SetContent("{\"name\":\"Alice\"}")
            .AddHeader("Authorization", "Bearer token123")
            .Build();

        Console.WriteLine(request);
    }
}

File: CreationalPattern/BuilderPattern/BuilderPattern.rb
# Simple builder pattern for constructing HTTP-style requests

class HttpRequest
  attr_reader :method, :url, :headers, :params, :body

  def initialize(method:, url:, headers: {}, params: {}, body: nil)
    @method = method
    @url = url
    @headers = headers
    @params = params
    @body = body
  end

  def to_s
    param_string = params.map { |k, v| "#{k}=#{v}" }.join("&")
    full_url = param_string.empty? ? url : "#{url}?#{param_string}"
    <<~REQ
    #{method} #{full_url}
    Headers:
    #{headers.map { |k, v| "  #{k}: #{v}" }.join("\n")}
    Body:
      #{body}
    REQ
  end
end

# Builder class
class HttpRequestBuilder
  def initialize(method, url)
    @method = method
    @url = url
    @headers = {}
    @params = {}
    @body = nil
  end

  def add_header(key, value)
    @headers[key] = value
    self
  end

  def add_param(key, value)
    @params[key] = value
    self
  end

  def set_body(body)
    @body = body
    self
  end

  def build
    HttpRequest.new(method: @method, url: @url, headers: @headers, params: @params, body: @body)
  end
end

# Usage
request = HttpRequestBuilder.new("POST", "https://api.example.com/users")
            .add_header("Authorization", "Bearer abc123")
            .add_param("version", "1.0")
            .set_body("{\"name\":\"Alice\"}")
            .build

puts request
